src/EnergyPlus/DataHeatBalance.cc

```diff
@@ -2451,6 +2452,46 @@ namespace DataHeatBalance {
         }
     }
 
+    Real64 ConstructionData::setUserTemperatureLocationPerpendicular(Real64 userValue)
+    {
+        if (userValue < 0.0) {
+            ShowWarningError("Construction:InternalSource has a perpendicular temperature location parameter that is less than zero.");
+            ShowContinueError("Construction=" + this->Name + " has this error.  The parameter has been reset to 0.");
+            return 0.0;
+        } else if (userValue > 1.0) {
+            ShowWarningError("Construction:InternalSource has a perpendicular temperature location parameter that is greater than one.");
+            ShowContinueError("Construction=" + this->Name + " has this error.  The parameter has been reset to 1.");
+            return 1.0;
+        } else {    // Valid value between 0 and 1
+            return userValue;
+        }
+    }
+
+    void ConstructionData::setNodeSourceAndUserTemp(int &sourceNodeLocation,
+                                  int &userTempNodeLocation,
+                                  Array1D_int Nodes,
+                                  int NumOfPerpendNodes)
+    {
+        sourceNodeLocation = 0;
+        userTempNodeLocation = 0;
+        if (!this->SourceSinkPresent) return;
```

Comment:
@Myoldmopar I thought of that route initially, but because sourceNodeLocation and userTempNodeLocation have to be defined as 0 when the !SourceSinkPresent, it seemed to make more sense to leave this stuff in the function.  I can move those assignments outside the function but I feel like the code isn't as clean then.  Given all that, do you still feel like I should skip this function if not a SourceSink construction?

Reply:
How often is this being called?  It's a bit nit-picky of me, but I'll just say that if this is called a lot, it's probably worth skipping the call entirely, and just setting sourceNodeLocation and userTempNodeLocation to zero outside of the function.  The only reason is that you would end up benefitting slightly from not having to setup the function call arguments.  However, I should add some caveats, as I understand them.  The allocations of the function call are all made on the stack, which is, well, fast.  You are returning void, so that's not causing a performance hit.  Your arguments, however, reveal a potential issue.  The third argument `Nodes` is not passed by reference.  I believe this is going to end up making a new copy of the Array1D_int -- though assembly would verify what's going on under the hood with the Objexx array.  In addition, I see your point about code cleanliness.  With all these random thoughts in mind, my new suggestion:
- Make this function accept an `Array1D_int & Nodes` as the third argument
- Leave it as it is otherwise

Reply:
@Myoldmopar Sounds good--I will make that change.  This isn't called very often by the way.  Just once per construction that shows up in the user's input file.  The CTF calculations are only done once so it's not like this is happening every time step or something like that.

Reply:
Changed this and it will show up in the next commit.